### PR TITLE
Interpret the Cancelation Error More Clearly and Respond with Specific Error Code and Fix small semantic tokens bug

### PR DESCRIPTION
**Pull request by akshar2401**
This pr does two things at a high level
1) Interprets a cancellation error on a lsp request more clearly and returns a specific error code as specified the lsp protocol.

2)  Fixes a very small semantic tokens bug that i introduced during lsp refactoring

### CODE REVIEW

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Handlers/SemanticTokens/BaseSemanticTokensLanguageServerOperationHandler.cs [hunk 1]
```diff
@@ -47,7 +47,7 @@ private Task<IEnumerable<ITokenTextSpan>> GetTokensAsync(LanguageServerOperation
                 var result = scope?.Check(getTokensContext.expression);
                 if (result == null)
                 {
-                    return Task.FromResult(Enumerable.Empty<ITokenTextSpan>());
+                    return Task.FromResult<IEnumerable<ITokenTextSpan>>(null);
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Handlers/SemanticTokens/BaseSemanticTokensLanguageServerOperationHandler.cs:50]: >null [](http://example.com/codeflow?start=72&length=4)

double checking? 0-length to null ...

Comment (MikeStall) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/Handlers/SemanticTokens/BaseSemanticTokensLanguageServerOperationHandler.cs:50]: oh, I see .. it simplifies code on line 125.

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageServiceProtocol/RedesignedLanguageServerTests/Nl2FxTests.cs [hunk 2]
```diff
@@ -34,6 +34,10 @@ internal class TestNL2FxHandler : NLHandler
 
         public bool Delay { get; set; } = false;
 
+        public int Nl2FxDelayTime { get; set; } = 100;
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageServiceProtocol/RedesignedLanguageServerTests/Nl2FxTests.cs:37]: nit: Can we keep to 1 field: 
"Promote" boolean Delay into a Timespan?

File: src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageServiceProtocol/RedesignedLanguageServerTests/CancelRequestHandlerTests.cs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+            var scope = engine.CreateEditorScope(symbols: symbols);
+            var scopeFactory = new TestPowerFxScopeFactory((string documentUri) => scope);
+            Init(new InitParams(scopeFactory: scopeFactory));
+            var nl2FxHandler = CreateAndConfigureNl2FxHandler();
+            nl2FxHandler.Delay = true;
+            nl2FxHandler.ThrowOnCancellation = true;
+            nl2FxHandler.Nl2FxDelayTime = 2000;
+            var payload = NL2FxMessageJson(documentUri);
+            using var source = new CancellationTokenSource();
+            source.CancelAfter(1200);
```
Comment (MikeStall) [inline src/tests/Microsoft.PowerFx.Interpreter.Tests/LanguageServiceProtocol/RedesignedLanguageServerTests/CancelRequestHandlerTests.cs:32]: >1200 [](http://example.com/codeflow?start=31&length=4)

Can we shrink this?

My concern is that we have 50k tests and counting, even 1sec/test adds up..

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/JsonRpc/LanguageServerOutputBuilder.cs [hunk 4]
```diff
@@ -158,5 +166,56 @@ public static void AddMethodNotFoundError(this LanguageServerOutputBuilder outpu
...
+            if (cancellationToken.IsCancellationRequested)
+            {
+                outputBuilder.AddRequestCancelledError(id, exception?.GetDetailedExceptionMessage());
+                return true;
+            }
+
+            if (exception is OperationCanceledException || 
+
+                // Downstream nl2fx or fx2nl dependencies might throw InvalidOperationException with message containing "canceled".
+                // Instead of throwing OperationCanceledException
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/JsonRpc/LanguageServerOutputBuilder.cs:210]: Can we fix these instead?  Checking message string is very brittle.

**Comment by LucGenetier** [comment]
✅ No public API change.

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs [hunk 5]
```diff
@@ -196,15 +196,30 @@ private async Task<LanguageServerOutputBuilder> OnDataReceivedAsyncInternal(Lang
                 await handler.HandleAsync(context, cancellationToken).ConfigureAwait(false);
                 return outputBuilder;
             }
+            catch (OperationCanceledException operationCanceledEx)
+            {
+                return HandleException(outputBuilder, input, cancellationToken, operationCanceledEx);
+            }
```
Comment (MikeStall) [inline src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs:202]: Does this block do anything?
It looks like a copy of lines 203..206, which is what we'd execute anyways if this block didn't exist.

**Review by MikeStall — DISMISSED** [review]
:shipit:

File: src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs (hunk 5 above)
Comment (akshar2401) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs:202]: oops, i didn't realize i left a dup block

Comment (akshar2401) [reply src/libraries/Microsoft.PowerFx.LanguageServerProtocol/LanguageServer/LanguageServer.cs]: Addressed

**Comment by LucGenetier** [comment]
✅ No public API change.

**Comment by LucGenetier** [comment]
✅ No public API change.

**Review by MikeStall — APPROVED** [review]

**Comment by LucGenetier** [comment]
✅ No public API change.

**Pull request merged**